File: BOJ/27_동적 계획법과 최단거리 역추적/9252_LCS2.py
A = input()
B = input()

dp = [[0] * 1001 for _ in range(1001)]

ans = 0
for i in range(len(A)) :
    for j in range(len(B)) :
        if A[i] == B[j] :
            dp[i+1][j+1] = dp[i][j] + 1
            ans = max(ans,dp[i+1][j+1])
        else :
            dp[i+1][j+1] = max(dp[i][j+1],dp[i+1][j])

ret = ""
la = len(A); lb = len(B)

while dp[la][lb] != 0 :
    if dp[la][lb] == dp[la-1][lb] : la -= 1
    elif dp[la][lb] == dp[la][lb-1] : lb -= 1
    else :
        ret = A[la-1] + ret
        la-=1; lb-=1

print(ans)
print(ret)

File: BOJ/27_동적 계획법과 최단거리 역추적/9252_LCS2.cpp
#include <cstdio>
#include <string>
#include <iostream>

#define MAXN 1000

using namespace std;

string A,B;
int dp[MAXN+1][MAXN+1];

int main() {

    cin >> A >> B;
    int a = A.length(), b = B.length();

    for(int i=0; i<a; i++) {
        for(int j=0; j<b; j++) {
            if(A[i] == B[j]) dp[i+1][j+1] = dp[i][j] + 1;
            else dp[i+1][j+1] = max(dp[i+1][j],dp[i][j+1]);
        }
    }
    string ans = "";
    while(dp[a][b] != 0) {
        if(dp[a][b] == dp[a-1][b]) a -= 1;
        else if(dp[a][b] == dp[a][b-1]) b -= 1;
        else {
            ans = A[a-1] + ans;
            a-=1; b-=1;
        }
    }
    cout << dp[A.length()][B.length()] << "\n" << ans << endl;
    return 0;
}